packages/logger/src/logger.ts

```diff
@@ -137,23 +166,102 @@ export class Logger {
     this._logHandler = val;
   }
 
+  /**
+   * The optional, additional, user-defined log handler for the Logger instance.
+   */
+  private _userLogHandler: LogHandler | null = null;
+  get userLogHandler(): LogHandler | null {
+    return this._userLogHandler;
+  }
+  set userLogHandler(val: LogHandler | null) {
+    this._userLogHandler = val;
+  }
+
   /**
    * The functions below are all based on the `console` interface
    */
 
   debug(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.DEBUG, ...args);
     this._logHandler(this, LogLevel.DEBUG, ...args);
   }
   log(...args: unknown[]): void {
+    this._userLogHandler &&
+      this._userLogHandler(this, LogLevel.VERBOSE, ...args);
     this._logHandler(this, LogLevel.VERBOSE, ...args);
   }
   info(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.INFO, ...args);
     this._logHandler(this, LogLevel.INFO, ...args);
   }
   warn(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.WARN, ...args);
     this._logHandler(this, LogLevel.WARN, ...args);
   }
   error(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.ERROR, ...args);
     this._logHandler(this, LogLevel.ERROR, ...args);
   }
 }
+
+export function setLogLevel(level: LogLevelString | LogLevel): void {
+  const newLevel = typeof level === 'string' ? levelStringToEnum[level] : level;
+  instances.forEach(inst => {
+    inst.logLevel = newLevel;
+  });
+}
+
+export function setUserLogHandler(
+  logCallback: LogCallback | null,
+  options?: LogOptions
+): void {
+  if (logCallback !== null && typeof logCallback !== 'function') {
+    throw new TypeError(
+      'First argument to `onLog` must be null or a function.'
+    );
+  }
+  for (const instance of instances) {
+    let threshhold = instance.logLevel;
+    if (options && options.level) {
+      threshhold = levelStringToEnum[options.level];
+    }
+    if (logCallback === null) {
+      instance.userLogHandler = null;
+    } else {
+      instance.userLogHandler = (
+        instance: Logger,
+        level: LogLevel,
+        ...args: unknown[]
+      ) => {
+        const message = args
+          .map(arg => {
+            if (arg == null) {
+              return null;
+            } else if (typeof arg === 'string') {
+              return arg;
+            } else if (typeof arg === 'number' || typeof arg === 'boolean') {
+              return arg.toString();
+            } else if (arg instanceof Error) {
+              return arg.message;
+            } else {
+              try {
+                return JSON.stringify(arg);
+              } catch (ignored) {
+                return null;
+              }
+            }
+          })
+          .filter(arg => arg)
+          .join(' ');
+        if (level >= threshhold) {
```

Comment:
`threshhold` should be the loglevel that's passed in `options`. If `options` doesn't have it, it should be the current `instance.logLevel`. 

on line [224](https://github.com/firebase/firebase-js-sdk/pull/2434/files#diff-6d4e0e31c88a6d05d9b0a1fc0c63338fR224), `threshhold` should be assigned with `null` instead of the `instance.logLevel` at that time.

Reply:
Unless I misunderstand, I think that's the current behavior.  It will be set to `instance.logLevel` but if `options.level` exists, that will override it.

Reply:
You don't want to fix the loglevel to the value of `instance.logLevel` at the point of callback registration as it could be changed later. You want to use the up to date value of `instance.logLevel` when the callback is invoked.

Reply:
Oh, got it, fixed.